### PR TITLE
Fix preloading of st models on has_many through associations

### LINKED ISSUE #11078: Eager loading bug with has_many :through and STI (Rails 4.0.0rc2)
When using eager loading on a `has_many :through` to a model using STI, the output SQL seems to be incorrect:

``` ruby
class Group < ActiveRecord::Base
  has_many :users
  has_many :articles, through: :users
end

class User < ActiveRecord::Base
  belongs_to :group
  has_many :articles
end

class Post < ActiveRecord::Base
  belongs_to :user
end

class Article < Post
end
```

Running

``` ruby
Group.includes(:articles).to_a
```

results in

```
ActiveRecord::StatementInvalid: SQLite3::SQLException: 
no such column: posts.type: SELECT "users".* FROM "users"  
WHERE "posts"."type" IN ('Article') AND "users"."group_id" 
IN (298486374, 980190962, 980190963)
```

Dummy app showing the problem with test:
https://github.com/balvig/bug/blob/master/test/models/group_test.rb

### PR DESCRIPTION
**Pull request by moktin**
This patch fixes #11078 by removing the type conditions from `through_scope`.
It is really inspired by #11082 and the comment from @neerajdotname.

### CODE REVIEW

**Comment by AlfonsoUceda** [comment]
@moktin I know this is an old PR, but it could be rebased?

**Comment by moktin** [comment]
@AlfonsoUceda done, but I can't promise it will be merge.

File: activerecord/lib/active_record/core.rb [hunk 1]
```diff
@@ -308,12 +308,22 @@ def relation
...
           if finder_needs_type_condition? && !ignore_default_scope?
-            relation.where(type_condition).create_with(inheritance_column.to_s => sti_name)
+            relation_with_type_condition
           else
-            relation
+            relation_without_type_condition
           end
         end
 
+        def relation_with_type_condition #:nodoc:
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/core.rb]: Could you make this method public and keep the `# :nodoc:` so we don't have to use `send`?

Comment (rafaelfranca) [inline activerecord/lib/active_record/core.rb]: Also we should move those methods to `inheritance.rb` file with `finder_needs_type_condition`

File: activerecord/lib/active_record/core.rb [hunk 2]
```diff
@@ -308,12 +308,22 @@ def relation
...
+          relation = relation_without_type_condition
+
+          relation.where(type_condition).create_with(inheritance_column.to_s => sti_name)
+        end
+
+        def relation_without_type_condition #:nodoc:
+          relation = Relation.create(self, arel_table, predicate_builder)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/core.rb]: this variable is not needed.

File: activerecord/lib/active_record/core.rb [hunk 3]
```diff
@@ -308,12 +308,22 @@ def relation
           relation = Relation.create(self, arel_table, predicate_builder)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/core.rb]: This line is not needed anymore.

**Review by rafaelfranca — CHANGES_REQUESTED** [review]

File: activerecord/lib/active_record/associations/preloader/through_association.rb [hunk 4]
```diff
@@ -87,7 +87,13 @@ def through_scope
...
               unless reflection_scope.where_clause.empty?
                 scope.includes_values = Array(reflection_scope.values[:includes] || options[:source])
-                scope.where_clause = reflection_scope.where_clause
+                base_wheres = if reflection.klass.finder_needs_type_condition?
+                                reflection.klass.send(:relation_with_type_condition).where_clause
+                              else
+                                ::ActiveRecord::Relation::WhereClause.empty
+                              end
+
+                scope.where_clause = reflection_scope.where_clause - base_wheres
```
Comment (kamipo) [inline activerecord/lib/active_record/associations/preloader/through_association.rb]: How about it?

```diff
--- a/activerecord/lib/active_record/associations/preloader/through_association.rb
+++ b/activerecord/lib/active_record/associations/preloader/through_association.rb
@@ -87,6 +87,9 @@ def through_scope
             else
               unless reflection_scope.where_clause.empty?
                 scope.includes_values = Array(reflection_scope.values[:includes] || options[:source])
+                if reflection.klass.finder_needs_type_condition?
+                  reflection_scope = reflection_scope.unscope(where: reflection_scope.klass.inheritance_column)
+                end
                 scope.where_clause = reflection_scope.where_clause
               end
```

`WhereClause#-` not always works properly unlike `WhereClause#except`.

File: activerecord/lib/active_record/core.rb [hunk 5]
```diff
@@ -305,12 +305,10 @@ def cached_find_by_statement(key, &block)
         end
 
         def relation
-          relation = Relation.create(self, arel_table, predicate_builder)
-
           if finder_needs_type_condition? && !ignore_default_scope?
-            relation.where(type_condition).create_with(inheritance_column.to_s => sti_name)
+            relation_with_type_condition
```
Comment (kamipo) [inline activerecord/lib/active_record/core.rb]: It is unneccesary changes anymore.

File: activerecord/lib/active_record/inheritance.rb [hunk 6]
```diff
@@ -84,6 +84,16 @@ def finder_needs_type_condition? #:nodoc:
         :true == (@finder_needs_type_condition ||= descends_from_active_record? ? :false : :true)
       end
 
+      def relation_with_type_condition #:nodoc:
+        relation = relation_without_type_condition
+
+        relation.where(type_condition).create_with(inheritance_column.to_s => sti_name)
```
Comment (kamipo) [inline activerecord/lib/active_record/inheritance.rb]: ditto.

File: activerecord/lib/active_record/relation/where_clause.rb [hunk 7]
```diff
@@ -17,6 +17,13 @@ def +(other)
         )
       end
 
+      def -(other)
+        WhereClause.new(
+          predicates - other.predicates,
+          binds - other.binds
```
Comment (kamipo) [inline activerecord/lib/active_record/relation/where_clause.rb]: ditto.

File: activerecord/lib/active_record/core.rb [hunk 8]
```diff
@@ -305,12 +305,10 @@ def cached_find_by_statement(key, &block)
...
 
         def relation
-          relation = Relation.create(self, arel_table, predicate_builder)
-
           if finder_needs_type_condition? && !ignore_default_scope?
-            relation.where(type_condition).create_with(inheritance_column.to_s => sti_name)
+            relation_without_type_condition.where(type_condition).create_with(inheritance_column.to_s => sti_name)
           else
-            relation
+            relation_without_type_condition
```
Comment (kamipo) [inline activerecord/lib/active_record/core.rb]: The changes to this file is unnecessary.

File: activerecord/lib/active_record/inheritance.rb [hunk 9]
```diff
@@ -84,6 +84,10 @@ def finder_needs_type_condition? #:nodoc:
         :true == (@finder_needs_type_condition ||= descends_from_active_record? ? :false : :true)
       end
 
+      def relation_without_type_condition #:nodoc:
+        Relation.create(self, arel_table, predicate_builder)
+      end
+
```
Comment (kamipo) [inline activerecord/lib/active_record/inheritance.rb]: ditto.

**Comment by mjansing** [comment]
Any updates on this issue?

**Pull request closed without merging**